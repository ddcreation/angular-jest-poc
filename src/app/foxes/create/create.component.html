<h1>Create fox</h1>
<form (ngSubmit)="onSubmit()" [formGroup]="form">
  <label for="fox-name">Enter fox name</label><br />
  <input id="fox-name" type="text" formControlName="name" /><br />
  <div *ngIf="form.invalid && form.touched" class="errors">
    <h2>Errors</h2>
    <ul>
      <li *ngFor="let error of form.controls.name.errors | keyvalue">
        {{ error.key }}
      </li>
    </ul>
  </div>
  <hr />
  <button type="submit" [disabled]="form.invalid">Add fox</button>
</form>
