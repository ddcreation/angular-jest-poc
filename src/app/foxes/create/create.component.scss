form {
  label {
    font-size: 1.5rem;
    color: lightsalmon;
  }

  input {
    font-size: 2rem;
  }

  button {
    font-size: 1.5rem;
    text-transform: uppercase;
    width: 100%;
    padding: 20px;
    text-align: center;
  }
}

.errors {
  color: red;
}
