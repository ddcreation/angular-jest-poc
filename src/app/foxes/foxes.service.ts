import { Injectable } from '@angular/core';

export interface Fox {
  id: number;
  name: string;
}

@Injectable({
  providedIn: 'root',
})
export class FoxesService {
  public foxes: Fox[] = [];

  public add(foxname: string): Fox {
    const newFox = {
      id: this.foxes.length + 1,
      name: foxname,
    };

    this.foxes.push(newFox);
    return newFox;
  }
}
