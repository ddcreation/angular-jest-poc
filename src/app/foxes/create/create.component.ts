import { Component, OnInit } from '@angular/core';
import { FormGroup, FormControl, Validators } from '@angular/forms';
import { Router } from '@angular/router';
import { FoxesService } from '../foxes.service';

@Component({
  selector: 'app-create',
  templateUrl: './create.component.html',
  styleUrls: ['./create.component.scss'],
})
export class CreateComponent implements OnInit {
  public form: FormGroup;

  constructor(private _foxes: FoxesService, private _router: Router) {}

  public onSubmit(): void {
    if (this.form.valid) {
      this._foxes.add(this.form.value.name);
      this._router.navigate(['/foxes']);
    }
  }

  ngOnInit(): void {
    this.form = new FormGroup({
      name: new FormControl('', [
        Validators.required,
        Validators.minLength(3),
        Validators.maxLength(10),
      ]),
    });
  }
}
