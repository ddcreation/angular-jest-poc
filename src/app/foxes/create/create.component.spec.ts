import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { CreateComponent } from './create.component';
import { ReactiveFormsModule } from '@angular/forms';
import { Router } from '@angular/router';
import { FoxesService } from '../foxes.service';
import { By } from '@angular/platform-browser';

const validFoxName = 'Wefox';

// Mock services:
const mockFoxes = {
  add: jest.fn((foxname) => ({ id: 0, name: foxname })),
};
const mockRouter = {
  navigate: jest.fn(),
};

describe('CreateComponent', () => {
  let component: CreateComponent;
  let fixture: ComponentFixture<CreateComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [ReactiveFormsModule],
      declarations: [CreateComponent],
      providers: [
        {
          provide: FoxesService,
          useValue: mockFoxes,
        },
        {
          provide: Router,
          useValue: mockRouter,
        },
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CreateComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  // Use snapshot of the form
  it('should init with empty form', () => {
    expect(component.form).toMatchSnapshot();
  });

  it('form should match rules', () => {
    // Required validator:
    component.form.controls.name.patchValue('');
    expect(component.form.invalid).toBe(true);
    expect(Object.keys(component.form.controls.name.errors).length).toBe(1);
    expect(Object.keys(component.form.controls.name.errors)[0]).toBe(
      'required'
    );

    // MinLength validator:
    component.form.controls.name.patchValue('aa');
    expect(component.form.invalid).toBe(true);
    expect(Object.keys(component.form.controls.name.errors).length).toBe(1);
    expect(Object.keys(component.form.controls.name.errors)[0]).toBe(
      'minlength'
    );

    // MaxLength validator:
    component.form.controls.name.patchValue('aaaaaaaaaaaaaa');
    expect(component.form.invalid).toBe(true);
    expect(Object.keys(component.form.controls.name.errors).length).toBe(1);
    expect(Object.keys(component.form.controls.name.errors)[0]).toBe(
      'maxlength'
    );

    // OK:
    component.form.controls.name.patchValue(validFoxName);
    expect(component.form.valid).toBe(true);
  });

  it('button should be disabled if form invalid', () => {
    // Disabled at the beginning:
    const button = fixture.debugElement.nativeElement.querySelector('button');
    expect(button.disabled).toBe(true);

    // Name OK = enabled
    component.form.patchValue({
      name: validFoxName,
    });
    fixture.detectChanges();
    expect(button.disabled).toBe(false);

    // Name KO = disabled
    component.form.patchValue({
      name: 'KO',
    });
    fixture.detectChanges();
    expect(button.disabled).toBe(true);
  });

  it('should submit valid form', () => {
    spyOn(component, 'onSubmit');

    const mouseClick = new MouseEvent('click');
    const button = fixture.debugElement.query(By.css('button'));

    // Disabled button does not submit:
    button.nativeElement.dispatchEvent(mouseClick);
    fixture.detectChanges();
    expect(component.onSubmit).not.toHaveBeenCalled();

    // Enabled button does submit:
    component.form.patchValue({ name: validFoxName });
    fixture.detectChanges();
    button.nativeElement.dispatchEvent(mouseClick);
    expect(component.onSubmit).toHaveBeenCalledTimes(1);
  });

  it('should add fox and redirect on valid submit', () => {
    // spyOn(mockFoxes, 'add');
    // spyOn(mockRouter, 'navigate');

    // Invalid form: no redirect
    component.onSubmit();
    expect(mockFoxes.add.mock.calls.length).toBe(0);
    expect(mockRouter.navigate.mock.calls.length).toBe(0);

    // Valid form: redirect
    component.form.patchValue({ name: validFoxName });
    component.onSubmit();
    expect(mockFoxes.add.mock.calls.length).toBe(1);
    expect(mockRouter.navigate.mock.calls.length).toBe(1);
  });
});
