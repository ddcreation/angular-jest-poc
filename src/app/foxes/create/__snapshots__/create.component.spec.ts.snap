// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateComponent should init with empty form 1`] = `
FormGroup {
  "_onCollectionChange": [Function],
  "_onDisabledChange": Array [],
  "asyncValidator": null,
  "controls": Object {
    "name": FormControl {
      "_onChange": Array [
        [Function],
      ],
      "_onCollectionChange": [Function],
      "_onDisabledChange": Array [
        [Function],
      ],
      "_parent": [Circular],
      "_pendingValue": "",
      "asyncValidator": null,
      "errors": Object {
        "required": true,
      },
      "pristine": true,
      "status": "INVALID",
      "statusChanges": EventEmitter {
        "__isAsync": false,
        "_isScalar": false,
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
      "touched": false,
      "validator": [Function],
      "value": "",
      "valueChanges": EventEmitter {
        "__isAsync": false,
        "_isScalar": false,
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
    },
  },
  "errors": null,
  "pristine": true,
  "status": "INVALID",
  "statusChanges": EventEmitter {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
  "touched": false,
  "validator": null,
  "value": Object {
    "name": "",
  },
  "valueChanges": EventEmitter {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
}
`;
